inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.h

```diff
@@ -29,27 +32,44 @@ class MKLDNNPadNode : public MKLDNNNode {
         SYMMETRIC = 3
     };
 
-    void padConstantOrEdge(const float *srcData, float* dstData,
-                           InferenceEngine::SizeVector srcDims, InferenceEngine::SizeVector dstDims,
-                           const bool isEdge = false);
-    void padReflectOrSymmetric(const float *srcData, float* dstData,
-                               InferenceEngine::SizeVector srcDims, InferenceEngine::SizeVector dstDims,
-                               const bool isSymmetric = false);
+    void padConstant();
+    template<typename T> void padConstantCommon();
+    void padConstantZero();
+    void padEdge();
+    void padReflectOrSymmetric(const bool isSymmetric = false);
+
+    void prepareParams();
 
-    size_t getWorkAmountDst() const;
+    inline void getDstIdx(const InferenceEngine::SizeVector& indexes, size_t& dstIdx) const;
+    inline uint8_t* getDataPtr(const MKLDNNMemory& memoryPtr) const;
 
     PadMode padMode = CONSTANT;
     float padValue = 0.f;
     std::vector<unsigned int> padsBegin;
     std::vector<unsigned int> padsEnd;
 
     struct {
+        InferenceEngine::SizeVector srcDims;
+        InferenceEngine::SizeVector dstDims;
         InferenceEngine::SizeVector srcODims;
         InferenceEngine::SizeVector srcStrides;
         InferenceEngine::SizeVector dstStrides;
-        size_t padPointsNum = 0;
-        InferenceEngine::SizeVector padDims;
+        InferenceEngine::SizeVector srcDimsForReflectOrSymmetric;
+        size_t nDimsForWork;
+        size_t workAmount;
+        size_t lastDstDim;
+        size_t shift;
+        uint8_t sizeData;
     } params;
+
+    template<typename T> friend struct PadConstantEmitter;
+};
+
+template<typename T>
```

Comment:
This struct is already forward declared, so it could be moved to the cpp.

Reply:
Discussed offline